velox/exec/MergingVectorOutput.cpp

```diff
@@ -0,0 +1,94 @@
+/*
+* Copyright (c) Facebook, Inc. and its affiliates.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "velox/exec/MergingVectorOutput.h"
+
+namespace facebook::velox::exec {
+
+MergingVectorOutput::MergingVectorOutput(
+    velox::memory::MemoryPool* pool,
+    int64_t preferredOutputBatchBytes,
+    int32_t preferredOutputBatchRows,
+    int64_t minOutputBatchBytes,
+    int32_t minOutputBatchRows) :
+      pool_(pool),
+      preferredOutputBatchBytes_(preferredOutputBatchBytes),
+      preferredOutputBatchRows_(preferredOutputBatchRows),
+      minOutputBatchBytes_(minOutputBatchBytes),
+      minOutputBatchRows_(minOutputBatchRows) {
+}
+
+RowVectorPtr MergingVectorOutput::getOutput() {
+  if (noMoreInput_) {
+    flush();
+  }
+  if (outputQueue_.size() > 0) {
+    RowVectorPtr res = outputQueue_.front();
+    outputQueue_.pop();
+    return res;
+  }
+  return nullptr;
+}
+
+void MergingVectorOutput::addVector(RowVectorPtr vector) {
+  int inputSize = vector->size();
+  if (inputSize == 0) {
+    return;
+  }
+
+  // Avoid memory copying for pages that are big enough
+  if (vector->estimateFlatSize() >= minOutputBatchBytes_ || inputSize >= minOutputBatchRows_) {
+    flush();
+    outputQueue_.push(vector);
+    return;
+  }
+
+  buffer(vector);
+}
+
+void MergingVectorOutput::buffer(RowVectorPtr vector) {
+  int inputSize = vector->size();
+  if (!bufferInputs_) {
+    bufferInputs_ = std::static_pointer_cast<RowVector>(
+        BaseVector::create(vector->type(), bufferRows_, pool_));
+  }
+
+  bufferInputs_->resize(bufferRows_ + inputSize);
+
+  std::vector<VectorPtr> childrens = vector->children();
+  for (auto i = 0; i < childrens.size(); i++) {
+    VectorPtr& vectorPtr = bufferInputs_->childAt(i);
+    vectorPtr->resize(bufferRows_ + inputSize);
```

Comment:
IIUC, `resize` may cause vector copy because previous capacity is smaller than new size, and this will happen each time when merging two vectors. cc @mbasmanova